lib/client/session.go

```diff
@@ -220,6 +246,169 @@ func (ns *NodeSession) createServerSession() (*ssh.Session, error) {
 	return sess, nil
 }
 
+// handleX11Forwarding handles x11 channel requests for the given server session.
+// If x11 forwarding is not requested by the client, or it is rejected by the server,
+// then x11 channel requests will be rejected.
+func (ns *NodeSession) handleX11Forwarding(ctx context.Context, sess *ssh.Session) error {
+	if !ns.nodeClient.TC.X11ForwardingEnabled {
+		return ns.rejectX11Channels(ctx)
+	}
+
+	if err := ns.setupX11Forwarding(ctx); err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := x11.RequestX11Forwarding(sess, ns.x11Display, ns.x11Proto, ns.x11FakeCookie); err != nil {
+		// If the x11 forwarding request fails, we must reject all x11 channel requests.
+		return ns.rejectX11Channels(ctx)
+	}
+
+	// Start listening for new x11 channel requests from the server
+	// and start x11 forwarding on those channels
+	err := ns.serveX11Channels(ctx, sess)
+	return trace.Wrap(err)
+}
+
+// setupX11Forwarding sets up x11 forwarding for the client session.
+func (ns *NodeSession) setupX11Forwarding(ctx context.Context) error {
+	var xauthEntry *x11.XAuthEntry
+	var err error
+
+	ns.x11Display = os.Getenv(x11.DisplayEnv)
+	if ns.x11Display == "" {
+		log.Info("X11 forwarding requested but $DISPLAY not set")
+		return trace.BadParameter("$DISPLAY not set")
+	}
+
+	// Validate the client given display.
+	for _, c := range ns.x11Display {
+		if !(unicode.IsLetter(c) || unicode.IsNumber(c) || c == ':' || c == '/' || c == '.' || c == '-' || c == '_') {
+			return trace.BadParameter("display contains invalid character %q", c)
+		}
+	}
+
+	if ns.nodeClient.TC.X11ForwardingTimeout != 0 {
+		ns.x11RefuseTime = time.Now().Add(time.Second * time.Duration(ns.nodeClient.TC.X11ForwardingTimeout))
+	}
+
+	log.Debug("creating an xauth cookie for x11 forwarding")
+	if xauthEntry == nil {
```

Comment:
`xauthEntry` is always going to be nil at this point, right? Can we reduce some of the  nested if statements here?